docs/core-concepts/params.md

```diff
@@ -0,0 +1,168 @@
+
+const monthParam = createParam((value, { invalid }) => {
+  if (months.includes(value as any)) {
+    return value
+  }
+
+  invalid(`Invalid month: ${value}`)
+})
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+    month: monthParam,
+  }),
+})
+```
+
+By default custom params will use `toString` when serializing. But you can use a Get/Set param to define how the param should be serialized and deserialized. For example we can make the month param capitalized when parsing and lowercase when serializing.
+
+```ts
+import { createParam } from '@kitbag/router'
+
+const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'] as const
+
+const monthParam = createParam({
+  get: (value, { invalid }) => {
+    const month = value.charAt(0).toUpperCase() + value.slice(1)
+
+    if (months.includes(month as any)) {
+      return month
+    }
+
+    invalid(`Invalid month: ${month}`)
+  },
+  set: (value) => {
+    return value.toLowerCase()
+  },
+})
+```
+
+## Default Values
+You can define a default value for a param type when creating it. This value will be used when the param is optional and not provided in the url. An optional param of type `String` will have the type `string | undefined` when being accessed. But when a default value is provided it will have the type `string`, even if the value is missing from the url.
+
+```ts {8}
+const monthParam = createParam({
+  get: (value, { invalid }) => {
+    ...
+  },
+  set: (value) => {
+    ...
+  },
+  defaultValue: 'january',
+})
+```
+
+You can also define a default value for a individual param using the `withDefault` utility.
+
+```ts {7}
+import { createRoute, path, withDefault } from '@kitbag/router'
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+    month: withDefault(monthParam, 'january'),
+  }),
+})
+```
+
+Or create a new param type with a default value from an existing param type.
+
+```ts {7}
+import { createParam, withDefault } from '@kitbag/router'
+
+const monthParam = createParam((value, { invalid }) => {
+  ...
+})
+
+const monthParamWithDefault = withDefault(monthParam, 'january')
+```
+
+## Query Params
+So far we have only looked at params in the `path` property. But params can also be used in the `query` property. When using params in the `query` the param goes where you expect the value to be in the url's search string.
+
+```ts {3}
+const events = createRoute({
+  name: 'events',
+  query: 'highlight=[?highlight]',
+})
+```
+
+The param name and the search key do not have to be the same. In this example the url search string would like like `?highlight=hello` and the param would have the value `hello`. But the param's name is `term`. 
```

Comment:
```suggestion
The param name and the search key do not have to be the same. In this example the url search string might be `?highlight=hello`, which has a `term` param with the value `hello`.
```

Reply:
you mention the param name is `term`, but I don't think that's true based on the example provided